tests/profiling/test_uwsgi.py

```diff
@@ -0,0 +1,168 @@
+    socket_name = tempfile.mktemp()
+    cmd = ["uwsgi", "--need-app", "--die-on-term", "--socket", socket_name, "--wsgi-file", uwsgi_app]
+
+    def _run_uwsgi(*args):
+        env = os.environ.copy()
+        if sys.version_info[0] == 2:
+            # On Python 2, it's impossible to import uwsgidecorators without this hack
+            env["PYTHONPATH"] = os.path.join(
+                os.environ.get("VIRTUAL_ENV", ""),
+                "lib",
+                "python%s.%s" % (sys.version_info[0], sys.version_info[1]),
+                "site-packages",
+            )
+
+        return subprocess.Popen(cmd + list(args), stdout=subprocess.PIPE, stderr=subprocess.STDOUT, env=env)
+
+    try:
+        yield _run_uwsgi
+    finally:
+        os.unlink(socket_name)
+
+
+def test_uwsgi_threads_disabled(uwsgi):
+    proc = uwsgi()
+    stdout, _ = proc.communicate()
+    assert proc.wait() != 0
+    assert b"ddtrace.internal.uwsgi.uWSGIConfigError: enable-threads option must be set to true" in stdout
+
+
+def test_uwsgi_threads_enabled(uwsgi, tmp_path, monkeypatch):
+    filename = str(tmp_path / "uwsgi.pprof")
+    monkeypatch.setenv("DD_PROFILING_OUTPUT_PPROF", filename)
+    proc = uwsgi("--enable-threads")
+    worker_pids = _get_worker_pids(proc.stdout, 1)
+    # Give some time to the process to actually startup
+    time.sleep(3)
+    proc.terminate()
+    assert proc.wait() == 30
+    for pid in worker_pids:
+        assert os.path.exists("%s.%d.1" % (filename, pid))
+
+
+def test_uwsgi_threads_processes_no_master(uwsgi, monkeypatch):
+    proc = uwsgi("--enable-threads", "--processes", "2")
+    stdout, _ = proc.communicate()
+    assert (
+        b"ddtrace.internal.uwsgi.uWSGIConfigError: master option must be enabled when multiple processes are used"
+        in stdout
+    )
+
+
+def _get_worker_pids(stdout, num_worker, num_app_started=1):
+    worker_pids = []
+    started = 0
+    while True:
+        line = stdout.readline()
+        if line == b"":
+            break
+        elif b"WSGI app 0 (mountpoint='') ready" in line:
+            started += 1
+        else:
+            m = re.match(r"^spawned uWSGI worker \d+ .*\(pid: (\d+),", line.decode())
+            if m:
+                worker_pids.append(int(m.group(1)))
+
+        if len(worker_pids) == num_worker and num_app_started == started:
+            break
+
+    return worker_pids
+
+
+def test_uwsgi_threads_processes_master(uwsgi, tmp_path, monkeypatch):
+    filename = str(tmp_path / "uwsgi.pprof")
+    monkeypatch.setenv("DD_PROFILING_OUTPUT_PPROF", filename)
+    proc = uwsgi("--enable-threads", "--master", "--processes", "2")
+    worker_pids = _get_worker_pids(proc.stdout, 2)
+    # Give some time to child to actually startup
+    time.sleep(3)
+    proc.terminate()
+    assert proc.wait() == 0
+    for pid in worker_pids:
+        assert os.path.exists("%s.%d.1" % (filename, pid))
+
+
+def test_uwsgi_threads_processes_master_lazy_apps(uwsgi, tmp_path, monkeypatch):
```

Comment:
These tests are basically just asserting that the profiler started up, successfully shutdown/flushed to disk, and didn't crash the app?

Reply:
That sounds like a very reductive way to resume my hours of fights with uWSGI, but yes, that's it.
😭

Reply:
Not a conversation that needs to happen here, but I am curious how we get tests like this not just for the profiler as well. Seems like something useful for a lot of different scenarios of uWSGI, or gunicorn, or mod_wsgi, or celery, or anything else like that.